### PR TITLE
Fix memory leak with ujson module

### LINKED ISSUE #49222: BUG: Reference leak in pandas/_libs/src/ujson/python/objToJSON.c
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
n/a
```

### Issue Description

Various code starting at https://github.com/pandas-dev/pandas/blob/890d09753492242bd30cc43c7ebd4e819ad89bf1/pandas/_libs/src/ujson/python/objToJSON.c#L154 calls `PyObject_GetAttrString`, but never DECREFs the values returned from the function. This means that these references are leaked, and the objects are never deallocated even if pandas is unimported.

I am happy to prepare a PR to fix this, but I'm not entirely sure what the code in question is even trying to do; it simply accesses these attributes and does nothing with the result.

Found via python/cpython#98253.

### Expected Behavior

No memory leaks

### Installed Versions

```
>>> pandas.show_versions()
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

INSTALLED VERSIONS
------------------
commit           : 91111fd99898d9dcaa6bf6bedb662db4108da6e6
python           : 3.11.0.candidate.2
python-bits      : 64
OS               : Darwin
OS-release       : 21.6.0
Version          : Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.5.1
numpy            : 1.23.4
pytz             : 2022.5
dateutil         : 2.8.2
setuptools       : 63.2.0
pip              : 22.3
Cython           : 0.29.32
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.9.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.5.0
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : 0.8.10
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None
```

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #49222 (Replace xxxx with the GitHub issue number)

This is mostly vendored from ujson upstream

https://github.com/ultrajson/ultrajson/blob/main/python/ujson.c

Using a static PyObject the way we did before is strongly discouraged. This is more verbose but should be more correct, although interestingly pushes us back to the legacy module initialization instead of PEP 489 multi phase initialization

### CODE REVIEW

**Comment by WillAyd** [comment]
@JelleZijlstra if you have the ability to confirm this patch fixes your issue would be appreciated

**Comment by akx** [comment]
As an aside, has Pandas considered https://github.com/ijl/orjson (a Rust-based JSON (de)serializer)?

**Review by mroeschke — COMMENTED** [review]
Thanks @WillAyd. Do you think it's worth adding and ASV `peakmem_*` benchmarks or are we sufficiently covered?

**Comment by WillAyd** [comment]
Its a pretty small memory leak. I don't think it would even register on a peakmem ASV. Let's see if it helps @jellezijlstra and can go from there

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, I can confirm this removes the leak. I left a few comments on the code.

@mroeschke this memleak only appears when pandas is unimported from a running process. It's unlikely to affect many current use cases.

File: pandas/_libs/src/ujson/python/ujson.c [hunk 1]
```diff
@@ -67,15 +67,386 @@ static PyMethodDef ujsonMethods[] = {
...
+        return NULL;
+    }
+
+    // PyModule_AddStringConstant(module, "__version__", UJSON_VERSION);
+
+#ifndef PYPY_VERSION
+    PyObject *mod_decimal = PyImport_ImportModule("decimal");
+    if (mod_decimal) {
+        PyObject *type_decimal = PyObject_GetAttrString(mod_decimal, "Decimal");
+        assert(type_decimal != NULL);
```
[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/ujson.c:394]: Yea it's a good question. I copied this directly from ultrajson upstream so wanted to stay as true to upstream as possible, but no problem changing

[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pandas/_libs/src/ujson/python/ujson.c:394]: Ah, didn't realize that. Makes sense to just keep it in line with upstream in that case.

**Comment by mroeschke** [comment]
Looks good @WillAyd. Would be good to have a whatsnew note for this then LGTM

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @WillAyd

**Comment by rhshadrach** [comment]
This patch may have induced a potential regression. Please check the links below. If any ASVs are parameterized, the combinations of parameters that a regression has been detected appear as subbullets. This is a partially automated message.

 - https://asv-runner.github.io/asv-collection/pandas/#io.json.ToJSON.time_to_json
   - orient='split'; frame='df_int_float_str'
   - orient='split'; frame='df_int_floats'
   - orient='split'; frame='df_td_int_ts'
   - orient='values'; frame='df_int_float_str'
   - orient='values'; frame='df_int_floats'
   - orient='values'; frame='df_td_int_ts'
 - https://asv-runner.github.io/asv-collection/pandas/#io.json.ToJSONWide.time_to_json
   - orient='split'; frame='df_int_floats'
   - orient='split'; frame='df_td_int_ts'
   - orient='values'; frame='df_int_float_str'
   - orient='values'; frame='df_int_floats'
   - orient='values'; frame='df_td_int_ts'

**Pull request merged**